web/frontend/src/components/slam.vue

```diff
@@ -1,193 +1,208 @@
- *     req.setName(name);
- *     rcLogConditionally(req);
- *     const chunks: Uint8Array[] = [];
- *     const getPointCloudMapStream = props.client.slamService.getPointCloudMapStream(req);
- *     getPointCloudMapStream.on('data', (res) => {
- *       const chunk = res.getPointCloudPcdChunk_asU8();
- *       chunks.push(chunk);
- *     });
- *     getPointCloudMapStream.on('status', (status) => {
- *       if (status.code !== 0) {
- *         const error = {
- *           message: status.details,
- *           code: status.code,
- *           metadata: status.metadata,
- *         };
- *         displayError(error);
- *       }
- *     });
- *     getPointCloudMapStream.on('end', (end) => {
- *       if (end === undefined || end.code !== 0) {
- *         // the error will be logged in the 'status' callback
- *         return;
- *       }
- *       pointcloud = concatArrayU8(chunks);
- *       pointCloudUpdateCount += 1;
- *     });
- *   });
- * };
- */
+const fetchSLAMMap = (name: string): Promise<Uint8Array> => {
+  return new Promise((resolve, reject) => {
+    const req = new slamApi.GetPointCloudMapStreamRequest();
+    req.setName(name);
+    rcLogConditionally(req);
+    const chunks: Uint8Array[] = [];
 
-/*
- * const fetchSLAMPose = (name: string) => {
- *   return nextTick(() => {
- *     const req = new slamApi.GetPositionNewRequest();
- *     req.setName(name);
- *     props.client.slamService.getPositionNew(req, new grpc.Metadata(), (error, res): void => {
- *       if (error) {
- *         displayError(error);
- *         return;
- *       }
- *       pose = res!.getPose()!;
- *     });
- *   });
- * };
- */
+    const getPointCloudMapStream = props.client.slamService.getPointCloudMapStream(req);
+    getPointCloudMapStream.on('data', (res) => {
+      const chunk = res.getPointCloudPcdChunk_asU8();
+      chunks.push(chunk);
+    });
+    getPointCloudMapStream.on('status', (status) => {
+      if (status.code !== 0) {
+        const error = {
+          message: status.details,
+          code: status.code,
+          metadata: status.metadata,
+        };
+        reject(error);
+      }
+    });
+    getPointCloudMapStream.on('end', (end) => {
+      if (end === undefined || end.code !== 0) {
+        // the error will be logged in the 'status' callback
+        return;
+      }
+      const arr = concatArrayU8(chunks);
+      resolve(arr);
+    });
+  });
+};
+
+const fetchSLAMPose = (name: string): Promise<commonApi.Pose> => {
+  return new Promise((resolve, reject): void => {
+    const req = new slamApi.GetPositionNewRequest();
+    req.setName(name);
+    props.client.slamService.getPositionNew(req, new grpc.Metadata(), (error, res): void => {
+      if (error) {
+        reject(error);
+        return;
+      }
+      resolve(res!.getPose()!);
+    });
+  });
+};
 
 const refresh2d = async (name: string) => {
+  const map = await fetchSLAMMap(name);
+  const returnedPose = await fetchSLAMPose(name);
```

Comment:
I'm unsure whether we want the pose or map call to happen first. As is, we will be getting the most up-to-date data on the pose (PRO), however, that pose might also be outside the map if the map generation takes too long and the robot has time to move into an unexplored region (CON).

Thoughts?

Reply:
Note: Micheal's comment below may solve this conundrum

Reply:
If we ask for the pose & then the map then fake slam will be broken. This is because requesting the map increments time in fake slam, so if we ask for the pose & then ask for the map, the pose is from an earlier point in time than the map.

Also, in real slam, the map takes much longer to respond than the pose, so I think we will want to defer requesting where the robot is until the last possible moment to provide the most accurate robot marker location.